parsers/UY.py

```
@@ -46,118 +62,182 @@ def get_salto_grande(session: Optional[Session]) -> float:
 
 
 def fetch_production(
     zone_key: str = "UY",
-    session: Optional[Session] = None,
+    session: Optional[Session] = Session(),
     target_datetime: Optional[datetime] = None,
     logger: Logger = getLogger(__name__),
-) -> dict:
-    if target_datetime:
-        raise NotImplementedError("This parser is not yet able to parse past dates")
+) -> Union[dict, List[dict]]:
 
-    obj = parse_page(session)
+    if target_datetime is not None:
+        raise ParserException("This parser is unnable to parse dates in the past")
 
-    data = {
-        "zoneKey": zone_key,
-        "datetime": obj["datetime"],
-        "production": dict(
-            [(k, obj[INV_MAP_GENERATION[k]]) for k in INV_MAP_GENERATION.keys()]
-        ),
-        "source": "ute.com.uy",
-    }
+    # make a helper function to create output format
+    def make_datum(entry):
+        return {
+            "zoneKey": zone_key,
+            "datetime": entry["time"].datetime,
+            "production": {key: entry[key] for key in AVALIABLE_KEYS},
+            "source": "ute.com.uy",
+        }
 
-    return data
+    # then delegate all the actual work to get_entry_list
+    return get_entry_list(session, make_datum, correct_hydro=True)
 
 
 def fetch_exchange(
     zone_key1: str = "UY",
     zone_key2: str = "BR-S",
-    session: Optional[Session] = None,
+    session: Optional[Session] = Session(),
     target_datetime: Optional[datetime] = None,
     logger: Logger = getLogger(__name__),
 ) -> dict:
     """Requests the last known power exchange (in MW) between two countries."""
 
-    session = session or Session()
-    if target_datetime:
-        raise NotImplementedError("This parser is not yet able to parse past dates")
+    if target_datetime is not None:
+        raise ParserException("This parser is unnable to parse dates in the past")
 
     # set comparison
     if {zone_key1, zone_key2} != {"UY", "BR"}:
         return None
```

it should raise an ParserException instead of returning None.

---

should be fixed in 950e3dad8ab32c0cea9d86c7fee46a794ccc36f5